alns/accept/tests/test_great_deluge.py

```diff
@@ -0,0 +1,141 @@
+
+    # The first candidate is rejected and increases the threshold
+    great_deluge(None, Zero(), Zero(), One())
+    assert_equal(great_deluge._threshold, 0.01)
+
+    # The second candidate is accepted and decreases the threshold
+    great_deluge(None, Zero(), Zero(), Zero())
+    assert_equal(great_deluge._threshold, 0.0099)
+
+
+def test_linear_accepts_below_threshold():
+    great_deluge = GreatDeluge(2.01, 0.5, method="linear")
+
+    # Initial threshold is set at 2.01, hence Two should be accepted
+    assert_(great_deluge(None, One(), Zero(), Two()))
+
+
+def test_linear_rejects_above_threshold():
+    great_deluge = GreatDeluge(1.01, 0.5, method="linear")
+
+    # Initial threshold is set at 1.01, hence Two should be rejected
+    assert_(not great_deluge(None, One(), Zero(), Two()))
+
+
+def test_linear_rejects_equal_threshold():
+    great_deluge = GreatDeluge(2, 0.5, method="linear")
+
+    # Initial threshold is set at 2, hence One should be rejected
+    assert_(not great_deluge(None, One(), Zero(), Two()))
+
+
+"""
+Tests for non-linear variant
+"""
+
+
+def test_non_linear_raise_zero_best():
+    """
+    Test if an error is raised when the initial solution has value 0 and
+    the non-linear variant is used.
+
+    Reason: the relative gap cannot be computed.
+    """
+    great_deluge = GreatDeluge(2, 0.1, 0.01, 1, method="non-linear")
+
+    with assert_raises(ValueError):
+        great_deluge(None, Zero(), One(), One())
+
+
+def test_non_linear_threshold_update():
+    """
+    Test if the threshold is correctly updated using the non-linear update
+    method.
+    """
+    great_deluge = GreatDeluge(1.5, 0.1, 0.02, 1, method="non-linear")
+
+    # The first candidate is rejected and linearly increases the threshold
+    great_deluge(None, One(), Zero(), Two())
+    assert_equal(great_deluge._threshold, 1.51)
+
+    # The second candidate is accepted and exponentially decreases the threshold
+    great_deluge(None, One(), Zero(), Zero())
+    new_threshold = 1.51 * np.exp(-1) + 1
+    assert_equal(great_deluge._threshold, new_threshold)
+
+
+def test_non_linear_accepts_below_threshold():
+    great_deluge = GreatDeluge(1.01, 0.5, 0.01, 1, method="non-linear")
+
+    # Initial threshold is set at 1.01, hence One should be accepted
+    assert_(great_deluge(None, One(), Zero(), One()))
+
+
+def test_non_linear_rejects_above_threshold():
+    great_deluge = GreatDeluge(2, 0.5, 1, 1, method="non-linear")
+
+    # Initial threshold is set at 2 and the current solution is Zero,
+    # hence Two should be rejected
+    assert_(not great_deluge(None, One(), Zero(), Two()))
+
+
+def test_non_linear_accepts_improving_current():
+    great_deluge = GreatDeluge(2, 0.5, 1, 1, method="non-linear")
+
+    # Two does not improve over the threshold (2) but does improve over the
+    # current solution value (3), hence One should be accepted
+    assert_(great_deluge(None, One(), Three(), Two()))
```

Comment:
Question: can we also test this without another mock state?

Reply:
Or maybe we should make one mock State with a variable objective:
```python
from numbers import Number


@dataclass
class VarObj(State):
    obj: Number

    def objective(self) -> Number:
        return self.obj
```
And then
```python
Zero = VarObj(0)
One = VarObj(1)
Two = VarObj(2)
```
for backwards compatiblity, but otherwise make whatever mock object that's needed available.

Reply:
> Question: can we also test this without the new mock state `Three`?

I don't think so: in non-linear great deluge, I can't use `Zero` because of division by zero. So that leaves me with two states, which are not enough states to test with.

> Or maybe we should make one mock State with a variable objective:
> 
> ```python
> from numbers import Number
> 
> 
> class VarObj(State):
>     """Test solution state object with variable objective."""
>     
>     def __init__(self, obj: Number):
>         self.obj = obj
> 
>     def objective(self) -> Number:
>         return self.obj
> ```
> 
> And then
> 
> ```python
> Zero = VarObj(0)
> One = VarObj(1)
> Two = VarObj(2)
> ```
> 
> for backwards compatiblity, but otherwise make whatever mock object that's needed available.

For backwards compatibility we need Zero, One, Two, to be callable right? How about we leave Zero, One, Two as is, and replace `Three` with `VarObj`?

Reply:
>For backwards compatibility we need Zero, One, Two, to be callable right? How about we leave Zero, One, Two as is, and replace Three with VarObj?

IIRC, they are just `alns.State` objects. Those have an `objective()` member method, but not much else. They are not required to be callable - other than via their `objective()` member.

Reply:
Sorry I misunderstood the proposed suggestion. 

If we rename all references of `Zero()` to `VarObj(0)` etc., do you feel that the code is easier to read? I'm a bit split about this one. Seeing `0` instead of `Zero` makes it easier to see immediately what is evaluated/compared, but there's also something "ugly" about seeing `VarObj` repeated multiple times.
```python
record_travel(rnd.RandomState(), VarObj(0), VarObj(0), VarObj(1))
```

Reply:
Oh! I see what you mean now.

In `states.py` we could do the following instead as drop-in replacements for now:
```python
Zero = lambda: VarObj(0)
One = lambda: VarObj(1)
Two: lambda: VarObj(2)
```
At some later point we can then simplify by taking out the lambdas, but I do not want to burden this PR with that change.